File: day20/part1.cpp
#include <iostream>
#include <math.h>
#include <stdlib.h>

using namespace std;

int main () {
	int minimum = 33100000;
	int gifts = 0;
	int house = 1;

	while (gifts < minimum) {
		gifts = 0;

		int squareRoot = (int) sqrt(house) + 1;
		for (int i = 1; i < squareRoot; i++) {
			if (house % i == 0) {
				gifts += 10 * i;
				if (house / i != i)
					gifts += 10 * house / i;
			}
		}

		house++;
	}

	cout << house - 1 << endl;
}

File: 2015/ruby/day20/part1.rb
#!/usr/bin/ruby

minimum = 33100000 / 10
house = 0
gifts = 0

while gifts < minimum
    house += 1
    elves = (1..Math.sqrt(house).to_i).select { |n| house % n == 0 }
    gifts = elves.inject(0){ |sum, elf| sum + (elf + house / elf) }
end

puts house
